### PR TITLE
[TPO] experimental TPO trainer

### PR DESCRIPTION
**Pull request by kashif**
# What does this PR do?

Fixes # (issue)

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

## AI writing disclosure

We welcome the use of AI tools to help with contributions. For transparency and to help us improve our review process, please indicate the level of AI involvement in this PR.

- [ ] No AI usage: the PR was written entirely by a human.
- [ ] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [ ] AI-generated: the PR was mostly or fully generated by an AI tool.

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag members/contributors who may be interested in your PR.

---

> [!NOTE]
> **Medium Risk**
> Introduces a new experimental training algorithm (`TPOTrainer`) with custom dataset preprocessing, loss computation, and batching logic; correctness issues could affect training stability/results. Risk is mitigated by added unit tests and docs, but it still touches core training-loop behavior for a new method.
> 
> **Overview**
> Adds an **experimental Triple Preference Optimization (TPO)** implementation via `experimental.tpo.TPOTrainer`/`TPOConfig`, including triple-preference dataset preprocessing (supports conversational + standard formats, implicit-prompt extraction with validation) and a new `DataCollatorForTriplePreference` that can optionally skip the `reference` branch when `tpo_alpha=0.0`.
> 
> Implements TPO loss variants (`sigmoid`, `hinge`, `ipo`, `tpo-l`) combining a chosen-vs-rejected contrastive objective with an optional gold `reference` NLL term, and logs new training metrics (entropy, token accuracy, implicit rewards/logps).
> 
> Adds docs + paper index entry, wires the new `tpo_trainer` page into the experimental docs nav, provides an end-to-end `examples/scripts/tpo.py` (including UltraFeedback-to-triple conversion), and introduces a comprehensive `tests/experimental/test_tpo_trainer.py` covering collator behavior, multiple loss types, conversational/implicit-prompt cases, `tpo_alpha=0.0`, missing/mismatched `reference`, and PEFT training.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit f06b4babd994735bb47e3b74229f894844423b64. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by kashif** [comment]
@codex review

File: docs/source/_toctree.yml [hunk 1]
```diff
@@ -135,4 +135,6 @@
     title: SSD
   - local: xpo_trainer
     title: XPO
+  - local: tpo_trainer
+    title: TPO
```
[the remark this answers is not in the document]
Comment (albertvillanova) [reply docs/source/_toctree.yml:139]: I agree with this: Experimental trainers are alphabetically sort.

File: trl/experimental/tpo/tpo_config.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+                - `"sigmoid"`: sigmoid loss from the original [TPO](https://huggingface.co/papers/2405.16681) paper.
+                - `"hinge"`: hinge loss on the normalized likelihood from the
+                  [SLiC](https://huggingface.co/papers/2305.10425) paper.
+                - `"ipo"`: IPO loss from the [IPO](https://huggingface.co/papers/2310.12036) paper.
+                - `"tpo-l"`: length-normalized TPO variant from the
+                  [TPO](https://huggingface.co/papers/2405.16681) paper, which adds a target reward margin
+                  `tpo_l_gamma` to the Bradley-Terry objective.
+
+        beta (`float`, *optional*, defaults to `0.1`):
```
Comment (albertvillanova) [inline trl/experimental/tpo/tpo_config.py]: Why not aligning with the default value in the paper? `paper_index.md` and `tpo.py` both explicitly use beta=0.01
```suggestion
        beta (`float`, *optional*, defaults to `0.01`):
```

File: trl/experimental/tpo/tpo_config.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    # Parameters that control the training
+    loss_type: str = field(
+        default="sigmoid",
+        metadata={
+            "help": "Type of loss to use.",
+            "choices": ["sigmoid", "hinge", "ipo", "tpo-l"],
+        },
+    )
+    beta: float = field(
+        default=0.1,
```
Comment (albertvillanova) [inline trl/experimental/tpo/tpo_config.py]:
```suggestion
        default=0.01,
```

File: docs/source/paper_index.md [hunk 4]
```diff
@@ -1403,6 +1403,44 @@ training_args = CPOConfig(
...
+Introduces Triple Preference Optimization (TPO), a preference learning method that aligns an LLM with three responses per prompt — a gold (`reference`) completion, a preferred (`chosen`) completion and a dispreferred (`rejected`) completion — in a single optimization step. TPO combines a contrastive objective on the (chosen, rejected) pair with a supervised NLL term on the gold response, removing the need for a separate SFT stage and the reference model used in DPO. Used in TRL via [`experimental.tpo.TPOTrainer`]. To reproduce the paper's setting (Llama-3-Base, 5K), use this configuration:
+
+```python
+from trl.experimental.tpo import TPOConfig
+
+training_args = TPOConfig(
+    loss_type="sigmoid",  # contrastive loss between chosen and rejected (Section 3 of the paper)
+    tpo_alpha=1.0,  # weight of the NLL term on the gold response (Section 3 of the paper)
+    beta=0.01,  # β temperature (Table 2 of the paper)
+    learning_rate=5e-7,  # Table 2 of the paper
```
Comment (albertvillanova) [inline docs/source/paper_index.md]: Isn't Table 6 instead?

```suggestion
    beta=0.01,  # β temperature (Table 6 of the paper)
    learning_rate=5e-7,  # Table 6 of the paper
```

File: docs/source/paper_index.md [hunk 5]
```diff
@@ -1403,6 +1403,44 @@ training_args = CPOConfig(
...
+To use the TPO-L variant (length-normalized log-probabilities with a target reward margin γ), set `loss_type="tpo-l"` and `tpo_l_gamma`:
+
+```python
+from trl.experimental.tpo import TPOConfig
+
+training_args = TPOConfig(
+    loss_type="tpo-l",  # length-normalized variant (Section 3 of the paper)
+    tpo_alpha=1.0,
+    beta=0.01,
+    tpo_l_gamma=5.4,  # γ target reward margin (Table 2 of the paper, Llama-3-Base 5K)
```
Comment (albertvillanova) [inline docs/source/paper_index.md]: The default value in the paper isn't 0.5? Also, isn't Table 6 instead?
```suggestion
    tpo_l_gamma=0.5,  # γ target reward margin (Table 6 of the paper, Llama-3-Base 5K)
```

File: trl/experimental/tpo/tpo.py [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+    --dataset_name tpo-alignment/triple-preference-ultrafeedback-40K \
+    --model_name_or_path Qwen/Qwen3-0.6B \
+    --per_device_train_batch_size 2 \
+    --max_steps 1000 \
+    --learning_rate 5e-7 \
+    --gradient_accumulation_steps 8 \
+    --beta 0.01 \
+    --tpo_alpha 1.0 \
+    --loss_type tpo-l \
+    --tpo_l_gamma 5.4 \
```
Comment (albertvillanova) [inline trl/experimental/tpo/tpo.py]:
```suggestion
    --tpo_l_gamma 0.5 \
```

File: trl/experimental/tpo/tpo.py [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+if __name__ == "__main__":
+    parser = HfArgumentParser((ScriptArguments, TPOConfig, ModelConfig))
+    script_args, training_args, model_args = parser.parse_args_into_dataclasses()
+
+    ################
+    # Model & Tokenizer
+    ################
+    model = AutoModelForCausalLM.from_pretrained(
+        model_args.model_name_or_path, trust_remote_code=model_args.trust_remote_code
+    )
```
Comment (albertvillanova) [inline trl/experimental/tpo/tpo.py:161]: When passing a model instance (instead of `str` model ID), the trainer bypasses "device_map" adjustment for distributed training.

To fix this:
- Either pass the model ID string, letting the trainer handle loading
- Or document that users need to handle device_map themselves

Comment (kashif) [reply trl/experimental/tpo/tpo.py:161]: good catch, now I mirror the model-loading pattern from `trl/scripts/dpo.py` using `get_quantization_config` +  `get_kbit_device_map` to set `device_map` correctly
